### PR TITLE
Detect package references containing dll's that were not used during compilation

### PR DESCRIPTION
**Pull request by pakrym**
Note: it does not tell you that you can safely remove package reference it just points out which dll's were not used during compilation. Package that has unused dll could have brought transitive reference that is used, content files, etc. 

Sometimes package references are unused only for one of the targets so users need to pay attention to which target produced the warning.

### CODE REVIEW

**Comment by pakrym** [comment]
Sample output:

```
C:\Users\pakrym\.nuget\packages\internal.aspnetcore.sdk\2.0.0-rc2-15234\build\FindUnusedReferences.targets(13,5): warning : Unused reference in Microsoft.Extensions.Configuration.Test.Common.csproj/net46 Microsoft.Extensions.FileProviders.Physical/2.0.0-preview1-24445/lib/netstandard1.3/Microsoft.Extensions.FileProviders.Physical.dll from package Microsoft.Extensions.FileProviders.Physical [D:\github\aspnet\Configuration\test\Microsoft.Extensions.Configuration.Test.Common\Microsoft.Extensions.Configuration.Test.Common.csproj]
  Microsoft.Extensions.Configuration.Test.Common -> D:\github\aspnet\Configuration\test\Microsoft.Extensions.Configuration.Test.Common\bin\Debug\net46\Microsoft.Extensions.Configuration.Test.Common.dll
C:\Users\pakrym\.nuget\packages\internal.aspnetcore.sdk\2.0.0-rc2-15234\build\FindUnusedReferences.targets(13,5): warning : Unused reference in Microsoft.Extensions.Configuration.AzureKeyVault.csproj/net46 System.Reflection.Metadata/1.4.2/lib/portable-net45+win8/System.Reflection.Metadata.dll from package System.Reflection.Metadata [D:\github\aspnet\Configuration\src\Microsoft.Extensions.Configuration.AzureKeyVault\Microsoft.Extensions.Configuration.AzureKeyVault.csproj]
```

File: src/Internal.AspNetCore.Sdk/build/FindUnusedReferences.targets [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+-->
+<Project>
+  <Target Name="FindUnusedReferences"
+          AfterTargets="CoreCompile"
+          DependsOnTargets="RunResolvePackageDependencies"
+          Condition=" '$(EnableFindUnusedReferences)' == 'true' ">
+    <Sdk_FindUnusedReferences Assembly="@(IntermediateAssembly)" References="@(ReferencePath)" Packages="@(PackageDependencies)" Files="@(FileDefinitions)" >
+      <Output TaskParameter="UnusedReferences" ItemName="UnusedReferences" />
+    </Sdk_FindUnusedReferences>
+    <Warning Condition="'@(UnusedReferences)' != ''" Text="Unused reference in $(MSBuildProjectFile)/$(TargetFramework) %(UnusedReferences.Identity) from package %(UnusedReferences.PackageName)" />
```
Comment (pranavkm) [inline src/Internal.AspNetCore.Sdk/build/FindUnusedReferences.targets]: Would it be possible to make this a verifier rule? We really don't have a good way to suppress individual warnings here and I rather not spend time every compilation doing extra work.

**Comment by natemcmaster** [comment]
How many warnings do you get when building Universe with this change? I'm concerned this will further clutter our build output with warnings that are not actionable.

I think with a few extra checks, this target could make strong recommendations about whether a package is safe to remove. Those additional checks might include:

 - listed as a PackageReference? 
 - build targets or content files actually in use? 
 - are its transitive references in use?

File: src/Internal.AspNetCore.Sdk/build/FindUnusedReferences.targets (hunk 1 above)
Comment (pakrym) [inline src/Internal.AspNetCore.Sdk/build/FindUnusedReferences.targets]: Not really, I need list of things that got passed to compiler and it's not easy to infer from assets.json anymore :(.

**Comment by pakrym** [comment]
@natemcmaster tons of warnings right now, this definitely needs to be disabled for tests and samples. Maybe this should be run only occasionally to cleanup libraries and not on every run, or it need a way to specify exclusions. 

1. It always looks only on direct package reference
2. I think that there are to many ways for package to be involved in build pipeline and it's to hard to track them all, my goal was not to make a tool that gives definite result but just points to a possible problem
3. Sometimes it's good to know that you can lower package reference to it's transitive dependency

**Comment by natemcmaster** [comment]
+1 reduce noise.

 - make it a diagnostic that only shows up if you run `/verbosity:Normal`. Most ppl won't care or won't know what to do with the info
- or, disable by default. Require setting `/p:FindUnusedReferences=true`

**Comment by kichalla** [comment]
> Sometimes it's good to know that you can lower package reference to it's transitive dependency

+1

Does this give warning in the scenario where I have a direct package reference but is not really required to be specified as its going to be brought transitively by another direct package reference that I have? (I feel VS should have built-in support for this)

**Comment by pakrym** [comment]
@kichalla no. It gives a warning when you reference a package but don't use it's dll, sometimes it happens that you at the same time use dll of transitive dependency in brings, so you can replace package reference with it's transitive dependency.

**Comment by pakrym** [comment]
This is disabled by default.

**Comment by pranavkm** [comment]
Is there a lot of utility in this if it's disabled by default?

**Comment by pakrym** [comment]
I will do https://github.com/aspnet/Coherence/issues/175 by enabling it per repo maybe fix more bugs and if we feel it gives useful results I'm going to enable it by default.

File: src/Internal.AspNetCore.BuildTools.Tasks/FindUnusedReferences.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            var references = new HashSet<string>(References.Select(item => item.ItemSpec), StringComparer.OrdinalIgnoreCase);
+            var referenceFiles = Files.Where(file => references.Contains(file.GetMetadata("ResolvedPath")))
+                .ToDictionary(item => Path.GetFileNameWithoutExtension(item.GetMetadata("ResolvedPath")), StringComparer.OrdinalIgnoreCase);
+
+            var directReferences = new HashSet<string>(
+                Packages.Where(p => string.IsNullOrEmpty(p.GetMetadata("ParentPackage"))).Select(i => i.ItemSpec),
+                StringComparer.OrdinalIgnoreCase);
+
+            using (var fileStream = File.OpenRead(Assembly))
+            using (PEReader reader = new PEReader(fileStream))
```
Comment (pranavkm) [inline src/Internal.AspNetCore.BuildTools.Tasks/FindUnusedReferences.cs]: `var` all day everyday.

**Pull request merged**